controllers/operatorconfigmap_controller.go

```diff
@@ -170,209 +172,249 @@ func (c *OperatorConfigMapReconciler) Reconcile(ctx context.Context, req ctrl.Re
-			return c.own(prometheusRule)
-		})
-		if err != nil {
-			c.log.Error(err, "failed to create/update prometheus rules")
-			return ctrl.Result{}, err
-		}
+			c.rbdDaemonSet = &appsv1.DaemonSet{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      csi.RBDDaemonSetName,
+					Namespace: c.OperatorNamespace,
+				},
+			}
+			err = c.createOrUpdate(c.rbdDaemonSet, func() error {
+				if err := c.own(c.rbdDaemonSet); err != nil {
+					return err
+				}
+				csi.SetRBDDaemonSetDesiredState(c.rbdDaemonSet)
+				return nil
+			})
+			if err != nil {
+				c.log.Error(err, "failed to create/update rbd daemonset")
+				return ctrl.Result{}, err
+			}
+
+			// Need to handle deletion of the csiDriver object, we cannot set
+			// ownerReference on it as its cluster scoped resource
+			cephfsCSIDriver := templates.CephFSCSIDriver.DeepCopy()
+			cephfsCSIDriver.ObjectMeta.Name = csi.GetCephFSDriverName()
+			if err := csi.CreateCSIDriver(c.ctx, c.Client, cephfsCSIDriver); err != nil {
+				c.log.Error(err, "unable to create cephfs CSIDriver")
+				return ctrl.Result{}, err
+			}
+
+			rbdCSIDriver := templates.RbdCSIDriver.DeepCopy()
+			rbdCSIDriver.ObjectMeta.Name = csi.GetRBDDriverName()
+			if err := csi.CreateCSIDriver(c.ctx, c.Client, rbdCSIDriver); err != nil {
+				c.log.Error(err, "unable to create rbd CSIDriver")
+				return ctrl.Result{}, err
+			}
+
+			prometheusRule := &monitoringv1.PrometheusRule{}
+			if err := k8sYAML.NewYAMLOrJSONDecoder(bytes.NewBufferString(string(pvcPrometheusRules)), 1000).Decode(prometheusRule); err != nil {
+				c.log.Error(err, "Unable to retrieve prometheus rules.", "prometheusRule", klog.KRef(prometheusRule.Namespace, prometheusRule.Name))
+				return ctrl.Result{}, err
+			}
+
+			prometheusRule.SetNamespace(c.OperatorNamespace)
+
+			err = c.createOrUpdate(prometheusRule, func() error {
+				applyLabels(c.operatorConfigMap.Data["OCS_METRICS_LABELS"], &prometheusRule.ObjectMeta)
+				return c.own(prometheusRule)
+			})
+			if err != nil {
+				c.log.Error(err, "failed to create/update prometheus rules")
+				return ctrl.Result{}, err
+			}
 
-		c.log.Info("prometheus rules deployed", "prometheusRule", klog.KRef(prometheusRule.Namespace, prometheusRule.Name))
+			c.log.Info("prometheus rules deployed", "prometheusRule", klog.KRef(prometheusRule.Namespace, prometheusRule.Name))
+		} else {
+			// deletion phase
+			result, err := c.deletionPhase()
+			if err != nil || !result.Requeue || result.RequeueAfter > 0 {
+				return result, err
+			}
+
+			//remove finalizer
+			controllerutil.RemoveFinalizer(c.operatorConfigMap, OperatorConfigMapFinalizer)
+			if err := c.Client.Update(c.ctx, c.operatorConfigMap); err != nil {
+				return ctrl.Result{}, err
+			}
+			c.log.Info("finallizer removed successfully")
+		}
 	}
+	return ctrl.Result{}, nil
+}
 
+func (c *OperatorConfigMapReconciler) deletionPhase() (reconcile.Result, error) {
```

Comment:
since the first return statement is always same can that be removed?

Reply:
not sure i'm following, can you please elaborate?

Reply:
You are always returning `ctrl.Result{}` in this func body and so that can be omitted right? Like,
```suggestion
func (c *OperatorConfigMapReconciler) deletionPhase() error {
```